zksync/sdk/musig/schnorr_musig_native.py

```diff
@@ -0,0 +1,176 @@
+import abc
+import ctypes.util
+from ctypes import Structure
+from ctypes import c_size_t
+from ctypes import c_ubyte
+from ctypes import c_uint32
+from enum import IntEnum
+
+LIBRARY_NAME = 'musig_c'
+STANDARD_ENCODING_LENGTH = 32
+AGG_SIG_ENCODING_LENGTH = 64
+
+
+class MusigRes(IntEnum):
+    OK = 0,
+    INVALID_INPUT_DATA = 1,
+    ENCODING_ERROR = 2,
+    SIGNATURE_VERIFICATION_FAILED = 3,
+    INTERNAL_ERROR = 4,
+    INVALID_PUBKEY_LENGTH = 100,
+    NONCE_COMMITMENT_NOT_GENERATED = 101,
+    NONCE_PRECOMMITMENTS_NOT_RECEIVED = 102,
+    NONCE_PRECOMMITMENTS_AND_PARTICIPANTS_NOT_MATCH = 103,
+    NONCE_COMMITMENTS_NOT_RECEIVED = 104,
+    NONCE_COMMITMENTS_AND_PARTICIPANTS_NOT_MATCH = 105,
+    SIGNATURE_SHARE_AND_PARTICIPANTS_NOT_MATCH = 106,
+    COMMITMENT_IS_NOT_IN_CORRECT_SUBGROUP = 106,
+    INVALID_COMMITMENT = 107,
+    INVALID_PUBLIC_KEY = 108,
+    INVALID_PARTICIPANT_POSITION = 109,
+    AGGREGATED_NONCE_COMMITMENT_NOT_COMPUTED = 110,
+    CHALLENGE_NOT_GENERATED = 111,
+    INVALID_SIGNATURE_SHARE = 112,
+    INVALID_SEED = 113
+
+
+class AggregatedPublicKey(Structure):
+    _fields_ = [
+        ("data", c_ubyte * STANDARD_ENCODING_LENGTH),
+    ]
+
+
+AggregatedPublicKeyPointer = ctypes.POINTER(AggregatedPublicKey)
+
+
+class MusigBN256Signer(Structure):
+    pass
+
+
+MusigBN256SignerPointer = ctypes.POINTER(MusigBN256Signer)
+
+
+class MusigSigner(Structure):
+    _fields_ = [
+        ("inner", MusigBN256SignerPointer),
+    ]
+
+
+MusigSignerPointer = ctypes.POINTER(MusigSigner)
+
+
+class Precommitment(Structure):
+    _fields_ = [
+        ("data", c_ubyte * STANDARD_ENCODING_LENGTH),
+    ]
+
+
+PrecommitmentPointer = ctypes.POINTER(Precommitment)
+
+
+class AggregatedCommitment(Structure):
+    _fields_ = [
+        ("data", c_ubyte * STANDARD_ENCODING_LENGTH),
+    ]
```

Comment:
Is there a reason why we can't represent all of the `Precommitment`, `Commitment` etc. with `bytes`? The current interface is not really user-friendly imo.

Reply:
My understanding is following.  We need to obey following requirements:
- we have technical limitation: there is a strict table of C type bindings https://docs.python.org/3/library/ctypes.html.; class ctypes.Structure can work only with `ctype.*` definitions.
- we have a strict contract from C library

We have a contract API in `schnorr_musig.h` and it defines `struct`:
```
typedef struct AggregatedPublicKey {
  uint8_t data[STANDARD_ENCODING_LENGTH];
} AggregatedPublicKey;
```
Structure uses uint8_t type. For sure, current type could be represented as ctypes.ubyte, ctypes.c_uint8, ctypes.c_char in python, because they have the same number of bits.

As a result, we are using array of ctypes.ubyte. However it is not possible to use python type `bytes` inside of 
 C definition `Structure`.

Conclusion:
- C bindings should not be usable. They should provide 1:1 matching for contract API. We have a strict contract.
- We are creating wrappers for usability. They may contain `bytes` definition.

In case of having extra questions, you can contact me directly.

Reply:
Okay, it is clear to me now why we need these structures on the low level, thank you.
In `schnor_musig.py`, can't we receive/return only `bytes`, converting them from/to these structures where necessary? This would greatly improve the development experience.

Reply:
All native classes was replaced by `bytes` type in `schnor_musig.py` and `schnorr_musig_signer.py` file.